src/rust/engine/src/nodes.rs

```diff
@@ -703,7 +703,8 @@ impl Snapshot {
   }
 
   fn store_path(item: &Path) -> Value {
-    externs::store_bytes(item.as_os_str().as_bytes())
+    // TODO(illicitonion): see https://github.com/pantsbuild/pants/issues/6302
+    externs::store_utf8(&item.to_string_lossy())
```

Comment:
If you add this to `externs.rs` just below `store_utf8`:

```
///
/// Store a buffer of utf8 bytes to pass to Python. This will end up as a Python `unicode`.
///
#[cfg(unix)]
pub fn store_utf8_osstr(utf8: &OsStr) -> Value {
  use std::os::unix::ffi::OsStrExt;
  let bytes = utf8.as_bytes();
  with_externs(|e| (e.store_utf8)(e.context, bytes.as_ptr(), bytes.len() as u64).into())
}
```

and here instead call `externs::store_utf8(item.as_os_str());`

you should be able to close https://github.com/pantsbuild/pants/issues/6302 :)